README.md
# form_builder_vat_field

VAT number input field for [flutter_form_builder](https://pub.dev/packages/flutter_form_builder) package

# Usage
```dart
FormBuilderVatField(
  name: 'vat_number',
  decoration: const InputDecoration(
    labelText: 'VAT Number',
    hintText: 'VAT number',
  ),
  priorityListByIsoCode: ['BE'],
  validator: FormBuilderValidators.compose([
    FormBuilderValidators.required(context),
    VatFormBuilderValidators.vatNumber(
      context,
      errorText: 'This is not a valid VAT number',
    ),
  ]),
  initialValue: 'FI 99999999',
),
```


CHANGELOG.md
## [0.1.2] - 15-March-2022
* Removed requirement validation

## [0.1.1] - 8-Sept-2021
* Updated dependency to Flutter Form Builder 6.1.0+1 and fixed code accordingly

## [0.1.0] - 25-May-2021
* Initial release.


test/vat_utils_test.dart
import 'package:test/test.dart';
import 'package:form_builder_vat_field/form_builder_vat_field.dart';
import 'dart:io';
import 'dart:convert';

Future<List<String>> readLinesFromFile(String filename) async {
  List<String> result = [];
  await File('test/$filename')
      .openRead()
      .map(utf8.decode)
      .transform(LineSplitter())
      .forEach((l) => result.add(l));
  return result;
}

void main() {
  test('Valid VAT numbers should be parsed correctly', () async {
    final validVatNumbers = await readLinesFromFile('valid.txt');

    for (final vatNumber in validVatNumbers) {
      expect(VatUtils.parse(vatNumber) != null, true,
          reason: '$vatNumber is not parsed as valid');
    }
  });

  test('Invalid VAT numbers should not be parsed correctly', () async {
    final invalidVatNumbers = await readLinesFromFile('invalid.txt');

    for (final vatNumber in invalidVatNumbers) {
      expect(VatUtils.parse(vatNumber) == null, true,
          reason: '$vatNumber is parsed as valid');
    }
  });
}


lib/src/vat_utils.dart
import 'vat_number.dart';

class VatUtils {
  static Map<String, String> _vatFormats = {
    'CH': r'^(CHE)([0-9]{9}(MWST|TVA|IVA))$',
    'AT': r'^(AT)(U[0-9]{8})$',
    'BE': r'^(BE)(0?[0-9]{9})$',
    'BG': r'^(BG)([0-9]{9,10})$',
    'CY': r'^(CY)([0-9]{8}[A-Z])$',
    'CZ': r'^(CZ)([0-9]{8,10})$',
    'DE': r'^(DE)([0-9]{9})$',
    'DK': r'^(DK)([0-9]{8})$',
    'EE': r'^(EE)([0-9]{9})$',
    'GR': r'^(EL|GR)([0-9]{9})$',
    'ES': r'^(ES)([0-9A-Z][0-9]{7}[0-9A-Z])$',
    'FI': r'^(FI)([0-9]{8})$',
    'FR': r'^(FR)([0-9A-Z]{2}[0-9]{9})$',
    'GB': r'^(GB)(([0-9]{9}([0-9]{3})?)|([A-Z]{2}[0-9]{3}))$',
    'HU': r'^(HU)([0-9]{8})$',
    'IE': r'^(IE)([0-9A-Z]{8,9})$',
    'IT': r'^(IT)([0-9]{11})$',
    'LT': r'^(LT)([0-9]{9}|[0-9]{12})$',
    'LU': r'^(LU)([0-9]{8})$',
    'LV': r'^(LV)([0-9]{11})$',
    'MT': r'^(MT)([0-9]{8})$',
    'NL': r'^(NL)([0-9]{9}B[0-9]{2})$',
    'PL': r'^(PL)([0-9]{10})$',
    'PT': r'^(PT)([0-9]{9})$',
    'RO': r'^(RO)([0-9]{2,10})$',
    'SE': r'^(SE)([0-9]{12})$',
    'SI': r'^(SI)([0-9]{8})$',
    'SK': r'^(SK)([0-9]{10})$',
    'HR': r'^(HR)([0-9]{11})$',
  };

  static List<String> get vatCountryCodes => _vatFormats.keys.toList();

  static VatNumber? parse(String vatNumber) {
    vatNumber = vatNumber
        .trim()
        .replaceAll(' ', '')
        .replaceAll('.', '')
        .replaceAll('-', '');

    for (final countryCode in _vatFormats.keys) {
      final pattern = _vatFormats[countryCode]!;
      final regex = RegExp(pattern, caseSensitive: false);
      final match = regex.firstMatch(vatNumber);
      if (match != null) {
        final countryCode = match.group(1);
        final number = match.group(2);
        if (countryCode != null && number != null) {
          var actualCountryCode = countryCode;
          if (countryCode == 'EL') {
            actualCountryCode = 'GR';
          } else if (countryCode == 'CHE') {
            actualCountryCode = 'CH';
          }
          return VatNumber(
            countryCode: actualCountryCode.toUpperCase(),
            nationalNumber: number.toUpperCase(),
          );
        } else {
          return null;
        }
      }
    }
    return null;
  }
}


lib/src/form_builder_vat_field.dart
import 'package:country_pickers/country.dart';
import 'package:country_pickers/country_pickers.dart';
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter/widgets.dart';
import 'package:flutter_form_builder/flutter_form_builder.dart';

import '../form_builder_vat_field.dart';
import 'vat_utils.dart';

/// Field for European VAT number input.
class FormBuilderVatField extends FormBuilderField<String> {
  final TextInputType keyboardType;
  final bool obscureText;
  final TextStyle? style;
  final TextEditingController? controller;
  final TextCapitalization textCapitalization;
  final TextInputAction? textInputAction;
  final StrutStyle? strutStyle;
  final TextDirection? textDirection;
  final TextAlign textAlign;
  final bool autofocus;
  final bool autocorrect;
  final MaxLengthEnforcement maxLengthEnforcement;
  final int? maxLength;
  final VoidCallback? onEditingComplete;
  final ValueChanged<String>? onFieldSubmitted;
  final List<TextInputFormatter>? inputFormatters;
  final double cursorWidth;
  final Radius? cursorRadius;
  final Color? cursorColor;
  final Brightness? keyboardAppearance;
  final EdgeInsets scrollPadding;
  final bool enableInteractiveSelection;
  final InputCounterWidgetBuilder? buildCounter;
  final bool expands;
  final int? minLines;
  final bool? showCursor;
  final VoidCallback? onTap;

  // For country dialog
  final String? searchText;
  final EdgeInsets? titlePadding;
  final bool? isSearchable;
  final Text? dialogTitle;
  final String? defaultSelectedCountryIsoCode;
  final List<String>? priorityListByIsoCode;
  final TextStyle? dialogTextStyle;
  final bool isCupertinoPicker;
  final double? cupertinoPickerSheetHeight;
  final TextAlignVertical? textAlignVertical;

  /// Creates field for European VAT number input.
  FormBuilderVatField({
    Key? key,
    //From Super
    required String name,
    FormFieldValidator<String>? validator,
    String? initialValue,
    InputDecoration decoration = const InputDecoration(),
    ValueChanged<String?>? onChanged,
    ValueTransformer<String>? valueTransformer,
    bool enabled = true,
    FormFieldSetter<String>? onSaved,
    AutovalidateMode autovalidateMode = AutovalidateMode.disabled,
    VoidCallback? onReset,
    FocusNode? focusNode,
    this.obscureText = false,
    this.textCapitalization = TextCapitalization.characters,
    this.scrollPadding = const EdgeInsets.all(20.0),
    this.enableInteractiveSelection = true,
    this.maxLengthEnforcement = MaxLengthEnforcement.enforced,
    this.textAlign = TextAlign.start,
    this.autofocus = false,
    this.autocorrect = true,
    this.cursorWidth = 2.0,
    this.keyboardType = TextInputType.text,
    this.style,
    this.controller,
    this.textInputAction,
    this.strutStyle,
    this.textDirection,
    this.maxLength,
    this.onEditingComplete,
    this.onFieldSubmitted,
    this.inputFormatters,
    this.cursorRadius,
    this.cursorColor,
    this.keyboardAppearance,
    this.buildCounter,
    this.expands = false,
    this.minLines,
    this.showCursor,
    this.onTap,
    this.searchText,
    this.titlePadding,
    this.dialogTitle,
    this.isSearchable,
    this.defaultSelectedCountryIsoCode = 'BE',
    this.priorityListByIsoCode,
    this.dialogTextStyle,
    this.isCupertinoPicker = false,
    this.cupertinoPickerSheetHeight,
    this.textAlignVertical,
  })  : assert(initialValue == null ||
            controller == null ||
            defaultSelectedCountryIsoCode != null),
        super(
          key: key,
          initialValue: initialValue,
          name: name,
          validator: validator,
          valueTransformer: valueTransformer,
          onChanged: onChanged,
          autovalidateMode: autovalidateMode,
          onSaved: onSaved,
          enabled: enabled,
          onReset: onReset,
          decoration: decoration,
          focusNode: focusNode,
          builder: (FormFieldState<String?> field) {
            final state = field as _FormBuilderVatFieldState;

            return InputDecorator(
              decoration: state.decoration,
              child: Row(
                children: <Widget>[
                  GestureDetector(
                    onTap: state.enabled
                        ? () {
                            state.requestFocus();
                            if (isCupertinoPicker) {
                              state._openCupertinoCountryPicker();
                            } else {
                              state._openCountryPickerDialog();
                            }
                          }
                        : null,
                    child: Row(
                      children: <Widget>[
                        const Icon(Icons.arrow_drop_down),
                        const SizedBox(width: 10),
                        CountryPickerUtils.getDefaultFlagImage(
                          state._selectedDialogCountry,
                        ),
                        const SizedBox(width: 10),
                        Text(
                          state._selectedDialogCountry.iso3Code == 'CHE'
                              ? 'CHE '
                              : '${state._selectedDialogCountry.isoCode} ',
                          style: Theme.of(state.context)
                              .textTheme
                              .subtitle1!
                              .merge(style),
                        ),
                      ],
                    ),
                  ),
                  Expanded(
                    child: TextField(
                      enabled: state.enabled,
                      style: style,
                      focusNode: state.effectiveFocusNode,
                      decoration: InputDecoration(
                        border: InputBorder.none,
                        enabledBorder: InputBorder.none,
                        errorBorder: InputBorder.none,
                        hintText: decoration.hintText,
                        hintStyle: decoration.hintStyle,
                      ),
                      onChanged: (val) {
                        state.invokeChange();
                      },
                      maxLines: 1,
                      keyboardType: keyboardType,
                      obscureText: obscureText,
                      onEditingComplete: onEditingComplete,
                      controller: state._effectiveController,
                      autocorrect: autocorrect,
                      autofocus: autofocus,
                      buildCounter: buildCounter,
                      cursorColor: cursorColor,
                      cursorRadius: cursorRadius,
                      cursorWidth: cursorWidth,
                      enableInteractiveSelection: enableInteractiveSelection,
                      maxLength: maxLength,
                      inputFormatters: inputFormatters,
                      keyboardAppearance: keyboardAppearance,
                      maxLengthEnforcement: maxLengthEnforcement,
                      scrollPadding: scrollPadding,
                      textAlign: textAlign,
                      textCapitalization: textCapitalization,
                      textDirection: textDirection,
                      textInputAction: textInputAction,
                      strutStyle: strutStyle,
                      //readOnly: state.readOnly, -- Does this need to be exposed?
                      expands: expands,
                      minLines: minLines,
                      showCursor: showCursor,
                      onTap: onTap,
                      textAlignVertical: textAlignVertical,
                    ),
                  ),
                ],
              ),
            );
          },
        );

  @override
  _FormBuilderVatFieldState createState() => _FormBuilderVatFieldState();
}

class _FormBuilderVatFieldState
    extends FormBuilderFieldState<FormBuilderVatField, String> {
  late TextEditingController _effectiveController;
  late Country _selectedDialogCountry;

  String get fullNumber {
    // When there is no VAT number text, the field is empty -- the country
    // prefix is only prepended when a VAT number is specified.
    final vatText = _effectiveController.text
        .trim()
        .replaceAll(' ', '')
        .replaceAll('.', '')
        .replaceAll('-', '');
    return vatText.isNotEmpty
        ? '${_selectedDialogCountry.iso3Code == 'CHE' ? 'CHE' : _selectedDialogCountry.isoCode}$vatText'
        : vatText;
  }

  @override
  void initState() {
    super.initState();
    _effectiveController = widget.controller ?? TextEditingController();
    _selectedDialogCountry = CountryPickerUtils.getCountryByIsoCode(
        widget.defaultSelectedCountryIsoCode!);
    _parseVat();
  }

  @override
  void dispose() {
    // Dispose the _effectiveController when initState created it
    if (null == widget.controller) {
      _effectiveController.dispose();
    }
    super.dispose();
  }

  /// cf. for reference: https://www.avalara.com/vatlive/en/eu-vat-rules/eu-vat-number-registration/eu-vat-number-formats.html
  Future<void> _parseVat() async {
    // print('initialValue: $initialValue');
    if (initialValue != null && initialValue!.isNotEmpty) {
      try {
        final parseResult = VatUtils.parse(initialValue!);
        if (parseResult != null) {
          setState(() {
            _selectedDialogCountry =
                CountryPickerUtils.getCountryByIsoCode(parseResult.countryCode);
          });
          _effectiveController.text = parseResult.nationalNumber;
        }
      } catch (error) {
        _effectiveController.text = initialValue!.replaceFirst('+', '');
      }
    }
  }

  void invokeChange() {
    didChange(fullNumber);
    widget.onChanged?.call(fullNumber);
  }

  void _openCupertinoCountryPicker() {
    showCupertinoModalPopup<void>(
      context: context,
      builder: (BuildContext context) {
        return CountryPickerCupertino(
          pickerSheetHeight: widget.cupertinoPickerSheetHeight ?? 300.0,
          onValuePicked: (Country country) {
            effectiveFocusNode!.requestFocus();
            setState(() => _selectedDialogCountry = country);
            didChange(fullNumber);
          },
          itemFilter: (c) => VatUtils.vatCountryCodes.contains(c.isoCode),
          priorityList: widget.priorityListByIsoCode != null
              ? List.generate(
                  widget.priorityListByIsoCode!.length,
                  (index) {
                    return CountryPickerUtils.getCountryByIsoCode(
                      widget.priorityListByIsoCode![index],
                    );
                  },
                )
              : null,
        );
      },
    );
  }

  void _openCountryPickerDialog() {
    showDialog<void>(
      context: context,
      builder: (context) {
        return Theme(
          data: Theme.of(context).copyWith(
            textSelectionTheme: TextSelectionThemeData(
              cursorColor: widget.cursorColor,
            ),
            primaryColor: widget.cursorColor ?? Theme.of(context).primaryColor,
          ),
          child: CountryPickerDialog(
            titlePadding: widget.titlePadding ?? const EdgeInsets.all(8.0),
            searchCursorColor:
                widget.cursorColor ?? Theme.of(context).primaryColor,
            searchInputDecoration:
                InputDecoration(hintText: widget.searchText ?? 'Search...'),
            isSearchable: widget.isSearchable ?? true,
            title: widget.dialogTitle ??
                Text(
                  'Select Your Country Code',
                  style: widget.dialogTextStyle ?? widget.style,
                ),
            onValuePicked: (Country country) {
              setState(() => _selectedDialogCountry = country);
              invokeChange();
            },
            itemFilter: (c) => VatUtils.vatCountryCodes.contains(c.isoCode),
            priorityList: widget.priorityListByIsoCode != null
                ? List.generate(
                    widget.priorityListByIsoCode!.length,
                    (index) {
                      return CountryPickerUtils.getCountryByIsoCode(
                          widget.priorityListByIsoCode![index]);
                    },
                  )
                : null,
            itemBuilder: _buildDialogItem,
          ),
        );
      },
    );
  }

  Widget _buildDialogItem(Country country) {
    return Container(
      child: ListTile(
        contentPadding: EdgeInsets.zero,
        leading: CountryPickerUtils.getDefaultFlagImage(country),
        title: Text(country.name),
        trailing:
            Text(country.iso3Code == 'CHE' ? 'CHE' : '${country.isoCode}'),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_form_builder/flutter_form_builder.dart';
import 'package:form_builder_vat_field/form_builder_vat_field.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  final _formKey = GlobalKey<FormBuilderState>();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text("FormBuilderVatField"),
      ),
      body: FormBuilder(
        key: _formKey,
        child: Padding(
          padding: const EdgeInsets.all(8.0),
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              FormBuilderVatField(
                name: 'vat_number',
                decoration: const InputDecoration(
                  labelText: 'VAT Number',
                  hintText: 'VAT number',
                ),
                // onChanged: _onChanged,
                priorityListByIsoCode: ['BE'],
                validator: FormBuilderValidators.compose([
                  FormBuilderValidators.required(context),
                  VatFormBuilderValidators.vatNumber(
                    context,
                    errorText: 'This is not a valid VAT number',
                  ),
                ]),
                initialValue: 'FI 99999999',
              ),
              const SizedBox(height: 15),
              ElevatedButton(
                onPressed: () {
                  if (_formKey.currentState!.saveAndValidate()) {
                    print(_formKey.currentState!.value);
                  }
                },
                child: Text("Submit"),
              )
            ],
          ),
        ),
      ),
    );
  }
}
